pkg/dynamicinstrumentation/ebpf/ebpf.go

```diff
@@ -0,0 +1,328 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+package ebpf
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/diagnostics"
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+)
+
+var (
+	bpffs = "/sys/fs/bpf" //TODO: Check via `mount(2)`
+)
+
+var (
+	globalTempDirPath string
+	globalHeadersPath string
+)
+
+func SetupRingBufferAndHeaders() {
+	tempDir, err := os.MkdirTemp("/tmp", "dd-go-di*")
+	if err != nil {
+		log.Info(err)
+	}
+
+	globalTempDirPath = tempDir
+
+	err = setupRingbufferAndHeaders()
+	if err != nil {
+		log.Errorf("couldn't establish bpf ringbuffer: %s", err)
+	}
+}
+
+func setupRingbufferAndHeaders() error {
+
+	// Create temporary directory structure to unload headers into
+	tempDir, err := os.MkdirTemp(globalTempDirPath, "run-*")
+	if err != nil {
+		return fmt.Errorf("couldn't create temp directory: %w", err)
+	}
+
+	headersPath, err := loadHeadersToTmpfs(tempDir)
+	if err != nil {
+		return fmt.Errorf("couldn't load headers to tmpfs: %w", err)
+	}
+
+	globalHeadersPath = headersPath
+
+	objFilePath := filepath.Join(globalTempDirPath, "ringbuffer-go-di.bpf.o")
+	cFlags := []string{
+		"-O2",
+		"-g",
+		"--target=bpf",
+		fmt.Sprintf("-I%s", globalHeadersPath),
+		"-o",
+		objFilePath,
+	}
+
+	// Read ringbuffer source file
+	ringbufferSource, err := os.ReadFile(filepath.Join(headersPath, "ringbuffer.h"))
+	if err != nil {
+		return fmt.Errorf("could not read ringbuffer source code: %w", err)
+	}
+
+	ringbufferSourcePath := filepath.Join(globalTempDirPath, "ringbuffer-go-di.bpf.c")
+
+	err = os.WriteFile(ringbufferSourcePath, ringbufferSource, 0644)
+	if err != nil {
+		return fmt.Errorf("could not write to ringbuffer source code file: %w", err)
+	}
+
+	// // Compile ringbuffer source file
+	// cfg := ddebpf.NewConfig()
+	// khOpts := kernel.HeaderOptions{
+	// 	DownloadEnabled: cfg.EnableKernelHeaderDownload,
+	// 	Dirs:            cfg.KernelHeadersDirs,
+	// 	DownloadDir:     cfg.KernelHeadersDownloadDir,
+	// 	AptConfigDir:    cfg.AptConfigDir,
+	// 	YumReposDir:     cfg.YumReposDir,
+	// 	ZypperReposDir:  cfg.ZypperReposDir,
+	// }
+	// kernelHeaders := kernel.GetKernelHeaders(khOpts, nil)
+	// if len(kernelHeaders) == 0 {
+	// 	return fmt.Errorf("unable to find kernel headers!")
+	// }
+
+	// err = compiler.CompileToObjectFile(ringbufferSourcePath, objFilePath, cFlags, []string{})
+	err = clang(cFlags, ringbufferSourcePath, withStdout(os.Stdout))
```

Comment:
@brycekahle Pointing this out as something to talk to you about. The commented out code produced the error that header files could not be found.